go/pools/fast_pool.go

```diff
@@ -0,0 +1,659 @@
+// to the pool.
+//
+// A SetCapacity of 0 is equivalent to closing the pool.
+func (p *FastPool) SetCapacity(capacity int, block bool) error {
+	err := p.setCapacityHandle(capacity)
+	if err != nil {
+		return err
+	}
+	if !p.State().Draining {
+		return nil
+	}
+
+	if block {
+		p.shrink()
+	} else {
+		go p.shrink()
+	}
+
+	return nil
+}
+
+// setCapacityHandle is used by SetCapacity to do sanity checks and change
+// state appropriately, while locking the state. It will also let the caller know
+// if the pool needs to shrink.
+func (p *FastPool) setCapacityHandle(capacity int) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.state.Closed {
+		return ErrClosed
+	}
+	if capacity < 0 || capacity > cap(p.pool) {
+		return fmt.Errorf("capacity %d is out of range", capacity)
+	}
+	if capacity > 0 {
+		minActive := p.state.MinActive
+		if capacity < minActive {
+			return fmt.Errorf("minActive %v would now be higher than capacity %v", minActive, capacity)
+		}
+	}
+	if capacity == 0 {
+		p.state.Closed = true
+	}
+
+	p.state.Draining = capacity < p.state.Capacity
+	p.state.Capacity = capacity
+	return nil
+}
+
+// shrink active resources until the active count is not above capacity.
+func (p *FastPool) shrink() {
+	for p.continueShrink() {
+		// We can't remove InUse resources, so only target the pool.
+		// Collect the InUse resources lazily when they're returned.
+		select {
+		case wrapper := <-p.pool:
+			wrapper.resource.Close()
+			wrapper.resource = nil
+
+			p.withLock(func() {
+				p.state.InPool--
+			})
+
+		case <-time.After(100 * time.Millisecond):
+			// Someone could have pulled from the pool just before
+			// we started waiting. Let's check the pool status again.
+		}
+	}
+}
+
+// continueShrink lets `shrink()` know to keep shrinking. It will also
+// set the pool state to closed when drained the capacity is 0.
+func (p *FastPool) continueShrink() bool {
+	p.Lock()
+	defer p.Unlock()
+
+	remaining := p.active() - p.state.Capacity
+	if remaining > 0 {
+		return true
+	}
+
+	p.state.Draining = false
+	if p.state.Capacity == 0 {
+		close(p.pool)
+		p.state.Closed = true
+	}
+	return false
+}
+
+// SetIdleTimeout sets the idle timeout for resources. The timeout is
+// checked at the 10th of the period of the timeout.
+func (p *FastPool) SetIdleTimeout(idleTimeout time.Duration) {
+	p.withLock(func() {
+		p.state.IdleTimeout = idleTimeout
+	})
+}
+
+// closeIdleResources scans the pool for idle resources
+// and closes them. It is meant to run as a goroutine and will
+// return when the pool is closed.
+func (p *FastPool) closeIdleResources() {
+	for !p.IsClosed() {
+
+		timeout := p.State().IdleTimeout
+
+		if timeout == 0 {
+			// Wait for an updated idleTimeout.
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
+		time.Sleep(timeout / 10)
+
+		var active, minActive int
+		var closed bool
+		for scanning, remaining := true, p.State().InPool; scanning && remaining > 0; remaining-- {
```

Comment:
Race condition: InPool is accessed without a lock here.

Reply:
I have moved InPool to be assigned a bit higher, with a description on how the loop handles changes to InPool after the initial read.